cvat/apps/engine/rq.py

```diff
@@ -0,0 +1,527 @@
+    RESULT = "result"
+    FUNCTION_ID = "function_id"
+    EXCEPTION_TYPE = "exc_type"
+    EXCEPTION_ARGS = "exc_args"
+    TMP_FILE = "tmp_file"
+
+
+class WithMeta(Protocol):
+    meta: dict[str, Any]
+
+
+class _AbstractRQMetaAttribute:
+    def __init__(
+        self, key: str, *, optional: bool = False, validator: Callable | None = None
+    ) -> None:
+        assert validator is None or callable(validator), "validator must be callable"
+        self._key = key
+        self._validator = validator
+        self._optional = optional
+
+
+class _GettableRQMetaAttribute(_AbstractRQMetaAttribute):
+    def __get__(self, instance: WithMeta, objtype: type[WithMeta] | None = None):
+        if self._optional:
+            return instance.meta.get(self._key)
+
+        return instance.meta[self._key]
+
+
+class _SettableRQMetaAttribute(_AbstractRQMetaAttribute):
+    def validate(self, value):
+        if value is None and not self._optional:
+            raise ValueError(f"{self._key} is required")
+        if value is not None and self._validator and not self._validator(value):
+            raise ValueError("Wrong type")
+
+    def __set__(self, instance: WithMeta, value: Any):
+        self.validate(value)
+        instance.meta[self._key] = value
+
+
+class ImmutableRQMetaAttribute(_GettableRQMetaAttribute):
+    pass
+
+
+class MutableRQMetaAttribute(_GettableRQMetaAttribute, _SettableRQMetaAttribute):
+    pass
+
+
+class UserRQMetaAttribute(ImmutableRQMetaAttribute):
+    def __init__(self, *, optional: bool = False, validator: Callable | None = None) -> None:
+        super().__init__(RQJobMetaField.USER, optional=optional, validator=validator)
+
+    def __get__(self, instance: WithMeta, objtype: type[WithMeta] | None = None):
+        assert RQJobMetaField.USER == self._key
+        return UserMeta(instance.meta[self._key])
+
+
+class RequestRQMetaAttribute(ImmutableRQMetaAttribute):
+    def __init__(self, *, optional: bool = False, validator: Callable | None = None) -> None:
+        super().__init__(RQJobMetaField.REQUEST, optional=optional, validator=validator)
+
+    def __get__(self, instance: WithMeta, objtype: type[WithMeta] | None = None):
+        assert RQJobMetaField.REQUEST == self._key
+        return RequestMeta(instance.meta[self._key])
+
+
+class UserMeta:
+    id: int = ImmutableRQMetaAttribute(
+        RQJobMetaField.UserField.ID, validator=lambda x: isinstance(x, int)
+    )
+    username: str = ImmutableRQMetaAttribute(
+        RQJobMetaField.UserField.USERNAME, validator=lambda x: isinstance(x, str)
+    )
+    email: str = ImmutableRQMetaAttribute(
+        RQJobMetaField.UserField.EMAIL, validator=lambda x: isinstance(x, str)
+    )
+
+    def __init__(self, meta: dict[RQJobMetaField.UserField, Any]) -> None:
+        self._meta = meta
+
+    @property
+    def meta(self) -> dict[RQJobMetaField.UserField, Any]:
+        return self._meta
+
+    def to_dict(self):
+        return self.meta
```

Comment:
Aren't these redundant?

Reply:
Okay, I also think so, but I made this way because:
1. `ImmutableRQMetaAttribute` expects objects with a `meta` attribute and I would not prefer to use `obj.to_dict()[SomeField]`
2. in some places we need to have the "dict" representation of an user or request (`BaseRQMeta.for_job(rq_job).user.to_dict()`). In such cases, I also don't like calling `BaseRQMeta.for_job(rq_job).user.meta` because it is less informative.
So, I kept `to_dict` only for `UserMeta` and `RequestMeta`.

Reply:
Okay, works for me.